tests/cpp/test_gpu_tensorcore.cpp

```diff
@@ -3222,6 +3222,224 @@ TEST_F(GPUTTensorCoreTest, MisalignedVectorization) {
+
+  FusionExecutor fe;
+  NVFUSER_TEST_CUDA_ARCH_COMPILE_CHECK(
+      8, 0, fe.compileFusion(&fusion, inputs, LaunchParams(), matmul_cparams));
+  ASSERT_TRUE(getBankConflictInfo(fe.kernel()).empty());
+  auto cg_outputs = fe.runFusion(inputs);
+  auto tref =
+      at::linear(A.to(at::kFloat), B.reshape({N1 * N2, K}).to(at::kFloat))
+          .reshape({M, N1, N2});
+  NVF_CHECK(cg_outputs[0].allclose(tref, 0.0001, 0.0001));
+}
+
+// Matmul test with multiple N dimensions that are discontiguous
+// TODO: This test currently fails, but it can be run using
+//   build/test_matmul --gtest_also_run_disabled_tests
+//
+// This case fails because the allocation domain of the B smem cached tensor is
+// [N1, K, N2], which is incompatible with ldMatrix. We need to gather the
+// allocation domains for the smem tensors by role instead, so that this
+// becomes [K, N1, N2].
+TEST_F(GPUTTensorCoreTest, DISABLED_MultipleNonConsecutiveNDims) {
+  int M = 504, N1 = 68, N2 = 2, K = 248;
+
+  Fusion fusion;
+  FusionGuard fg(&fusion);
+
+  // Note that N2 is inside K, so this is a TT layout
+  auto tv0 = makeContigTensor(2, DataType::Half); // M, K
+  auto tv1 = makeContigTensor(3, DataType::Half); // N1, K, N2
+
+  fusion.addInput(tv0);
+  fusion.addInput(tv1);
+
+  tv0 = broadcast(tv0, {false, true, false, true});
+  tv1 = broadcast(tv1, {true, false, false, false});
+
+  auto tv2 = fusedMultiplySum(tv0, tv1, {-2});
+
+  fusion.addOutput(tv2);
+
+  MatMulTileOptions gemm_tile;
+  gemm_tile.cta_tile = GemmTile(128, 128, 32);
+  gemm_tile.warp_tile = GemmTile(64, 64, 32);
+  gemm_tile.instruction_tile = GemmTile(16, 8, 16);
+
+  MatmulParams params;
+  // Output is M, N1, N2, contiguous so fully vectorizable despite size N2=2 in
+  // inner dim.
+  params.supported_vec_size = {8, 2, 4};
+  params.mma_macro = MmaMacro::Ampere_16_8_16;
+  params.tile_sizes = gemm_tile;
+  params.async_gmem_load_operands = true;
+  params.double_buffer_options.double_buffer_smem_write = true;
+  params.double_buffer_options.double_buffer_smem_read = true;
+  params.double_buffer_options.smem_double_buffer_stage = 4;
+  scheduleMatmul(&fusion, params);
+
+  auto options = at::TensorOptions().dtype(at::kHalf).device(at::kCUDA, 0);
+  at::Tensor A = at::randn({M, K}, options);
+  at::Tensor B = at::randn({N1, K, N2}, options);
+  std::vector<c10::IValue> inputs{A, B};
+
+  FusionExecutor fe;
+  NVFUSER_TEST_CUDA_ARCH_COMPILE_CHECK(
+      8, 0, fe.compileFusion(&fusion, inputs, LaunchParams(), matmul_cparams));
+  ASSERT_TRUE(getBankConflictInfo(fe.kernel()).empty());
+  auto cg_outputs = fe.runFusion(inputs);
+  auto Bpermuted = B.permute({{1, 2}}).reshape({N1 * N2, K});
+  auto tref = at::linear(A.to(at::kFloat), Bpermuted.to(at::kFloat))
+                  .reshape({M, N1, N2});
+  NVF_CHECK(cg_outputs[0].allclose(tref, 0.0001, 0.0001));
+}
+
+// This is a tougher test where we insert a batch dim between the M dims
+TEST_F(GPUTTensorCoreTest, MultipleMDimsBatch) {
```

Comment:
A batch dim can be inserted between M dims, but not the K dim. This is because the batch dimension gets parallelized, so that M1 and M2 will be contiguous with one another in the smem tensor.